test/service_test.cc

```diff
@@ -0,0 +1,171 @@
+#include <grpc++/grpc++.h>
+
+#include <chrono>
+
+#include "nighthawk/common/exception.h"
+
+#include "client/service_impl.h"
+
+#include "test/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/environment.h"
+#include "test/test_common/network_utility.h"
+
+#include "api/client/service.pb.h"
+#include "gtest/gtest.h"
+
+using namespace std::chrono_literals;
+using namespace testing;
+using testing::MatchesRegex;
+
+namespace Nighthawk {
+namespace Client {
+
+class ServiceTest : public TestWithParam<Envoy::Network::Address::IpVersion> {
+public:
+  void SetUp() override {
+    grpc::ServerBuilder builder;
+    int grpc_server_port = 0;
+    const std::string loopback_address =
+        Envoy::Network::Test::getLoopbackAddressUrlString(GetParam());
+
+    builder.AddListeningPort(fmt::format("{}:0", loopback_address),
+                             grpc::InsecureServerCredentials(), &grpc_server_port);
+    builder.RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    channel_ = grpc::CreateChannel(fmt::format("{}:{}", loopback_address, grpc_server_port),
+                                   grpc::InsecureChannelCredentials());
+    stub_ = std::make_unique<nighthawk::client::NighthawkService::Stub>(channel_);
+    setBasicRequestOptions();
+  }
+
+  void TearDown() override { server_->Shutdown(); }
+
+  void setBasicRequestOptions() {
+    auto options = request_.mutable_start_request()->mutable_options();
+    options->set_uri("http://127.0.0.1:10001/");
```

Comment:
Does this actually result in any loopback traffic being sent by NH? Or is this somehow mocked? I can't easily spot the later.

Reply:
It's fine in a followup (maybe add a TODO), thanks. My main concern is in IPv6-only setups, which we need to support at Google.

Reply:
If you can add a `TODO` we can resolve this one.